File: sort/sort.py
from random import randint

def selectionSort(ls):
	ls= ls[:]
	menor = 0;indexMenor = 0

	for i in range(0, len(ls)):
		for index in range(i, n):
			valor = ls[index]
			if(valor < menor or index == i):
				menor = valor
				indexMenor = index

		iAnterior= ls[i]
		ls[i] = ls[indexMenor]
		ls[indexMenor] = iAnterior

	return ls

def insertSort(ls):
	ls = ls[:]
	for i in range(1, len(ls)):
		elem = ls[i]
		ls.pop(i)
		for desci in range(i-1,-1,-1):
			if(ls[desci] < elem):
				ls.insert(desci+1, elem)
				break
			if(desci == 0):
				ls.insert(0, elem)

	return ls

def dividiOrdena(ls):
	medio = int(len(ls)/2)
	ls1 = ls[:medio]
	ls2 = ls[medio:]
	
	if(len(ls1) != 1):
		ls1 = dividiOrdena(ls1)
		ls2 = dividiOrdena(ls2)
		c = []
		i1 = 0
		i2 = 0
		lsComp = False; t2= len(ls2)
		while i1<len(ls1):
			if( lsComp or ls1[i1] <  ls2[i2] ):
				c.append(ls1[i1])
				i1+=1
			else:
				c.append(ls2[i2])
				if(i2+1 == t2):
					lsComp = True
				else:
					i2+=1
		if(t2+len(ls1) != len(c)):
			c+= ls2[i2:]
		return c
	else:
		if(ls[0]<ls[1]):
			return ls
		return [ls[1],ls[0]]


def mergeSort(ls):
	t = len(ls)
	if(t!=1):
		ls=ls[:]
		n = 2
		while True:
			if(n == t):
				break
			if(n > t):
				r = n-t
				for i in range(r):
					ls.insert(0, 0)
				break
			n*=2
		ls = dividiOrdena(ls)
		for i in range(r):
			ls.remove(0)

	return ls



def quickSort(ls):
	if(ls != []):
		ls = ls[:]
		t = len(ls)
		pivo = ls[t-1]
		ls.pop(t-1)
		if(t > 1):
			menor = []
			maior = []
			ultimoMaior = 0
			for i in ls:
				if(i <= pivo):
					menor +=[i]
				else:
					maior +=[i]

			if(menor != []):
				menor = quickSort(menor)
			if(maior != []):
				maior = quickSort(maior)

			ls = menor + [pivo] +maior

			return ls
		else:
			ls += [pivo]
			return ls
	else:
		return ls
	
lista= []
n = 20

for t in range(0,n):
    lista.append(randint(0, 50))

print("lista inicial: ",lista)


print("lista ordenada Selection sort: ", selectionSort(lista))
print("lista ordenada Insert sort: ", insertSort(lista))
print("lista ordenada Merge sort: ", selectionSort(lista))
print("Lista Ordenada Quick sort: ", quickSort(lista))

File: sort/sort.c
#include "stdio.h"
#include "stdlib.h"
#include "time.h"

typedef unsigned char BYTE;

void printArray(int* a, size_t size)
{
    int i;
    printf("[ %d", a[0]);
    for (i = 1; i < size; i++)
    {
        printf(", %d", a[i]);
    }
    printf(" ]\n");
}

void swapElement(void* a, int i1, int i2, size_t n){
    size_t i;
    BYTE temp;
    BYTE* ca;
    ca = (BYTE *)a;
    i1 = i1*n; i2 = i2*n;

    for(i=0; i < n; i++) {
        temp = ca[i1+i];
        ca[i1+i] = ca[i2+i];
        ca[i2+i] = temp;
    }

}

void bubbleSort(int* a, size_t size)
{
    int i, j;
    for (i = 0; i < size; i++){
        // printArray(a,size);
        for (j = 0; j < size-i-1; j++)
        {
            if (a[j] > *(a+j+1))
            {
                swapElement(a, j, j+1, sizeof(a[0]));
            }
        }
    }   
}

void selectionSort(int* a, size_t size)
{
    int i, j, lower;
    for(i = 0; i < size; i++){
        lower = i;
        for(j=i+1; j < size; j++){
            if(*(a+lower) > a[j]){
                lower = j;
            }
        }

        swapElement(a, i, lower, sizeof(a[0]));
    }

}

void insertSort(int* a, size_t size)
{
    int i, j, temp;

    for(i=1; i< size; i++){
        temp = a[i];
        for(j=i-1; j >= 0; j--){
            if(temp < *(a+j)){
                a[j+1] = a[j];
            }else{
                break;
            }
        }
        a[j+1] = temp;
    }
}

void quickSort(int* a, size_t size)
{
    int i, j, pivot;
    pivot = a[size-1];

    j = 0;
    for(i = 0; i < size-1;i++)
    {
        if(a[i] < pivot){
            swapElement(a, j, i, sizeof(int));
            j++;
        }
    }

    a[size-1] = a[j];
    a[j] = pivot;

    if(j) quickSort(&a[0], j);
    if(j != size-1) quickSort(&a[j+1], size-j-1);

} 

void mergeSort(int* a, size_t size)
{
    size_t i, j, k, mid;
    if(size > 2){
        mid = (size/2)+(size%2);
        mergeSort(&a[0], mid);
        mergeSort(&a[mid], size-mid);
        int la[mid];
        i = j =0;
        k = mid;
        for(i=0; i < mid; i++)
        {
            la[i] = a[i];
        }
        j = mid;
        i = k = 0;
        while (1)
        {
            if(k < mid && j < size){
                if(la[k] > a[j]){
                    a[i] = a[j];
                    j++;
                }else{
                    a[i] = la[k];
                    k++;
                }
                i++;
            }else if(k < mid){
                a[i] = la[k];
                k++;
                i++;
            }else{
                break;
            }
        }

        // for(i=mid; i < size;i++)
        // {
        //     for(j; j < i; j++){
        //         if(a[i] < a[j]){
        //             temp = a[i];
        //             for(x=i; x > j; x--) {a[x] = a[x-1];}
        //             a[j] = temp;   
        //             j++;
        //             break;
        //         }
        //     }
        // }

    }else if(size > 1){
        if(*(a) > *(a+1)){
            swapElement(a, 0, 1, sizeof(int));
        }
    }
}


int main()
{
    clock_t start, end;
    double delta;

    srand(time(NULL));

    size_t size = 70;
    int a1[size], a2[size], a3[size], a4[size], a5[size], i, n;
    for(i = 0; i < size; i++){
        n= (rand()%100)+1;
        a1[i] = n;
        a2[i] = n;
        a3[i] = n;
        a4[i] = n;
        a5[i] = n;
    }

    printf("Start array: ");
    printArray(a1, size);

    bubbleSort(a1, size);
    printf("Bubble sort: ");
    printArray(a1, size);

    selectionSort(a2, size);
    printf("Selection sort: ");
    printArray(a2, size);

    insertSort(a3, size);
    printf("Insert sort: ");
    printArray(a3, size);

    quickSort(a4, size);
    printf("Quick sort: ");
    printArray(a4, size);

    mergeSort(a5, size);
    printf("Merge sort: ");    
    printArray(a5, size);


    for(int i = 0; i < size; i++)
    {
        if(a4[i] != a5[i] || a1[i] != a2[i] || a2[i] != a3[i] || a3[i] != a4[i]){
            printf("BS: %d, SS: %d, IS: %d, QS: %d, MS: %d ", a1[i], a2[i], a3[i], a4[i], a5[i]);
            break;
        }       
    }

    printf("END");

    return 0;
}